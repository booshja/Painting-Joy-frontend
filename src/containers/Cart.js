// dependencies
import React, { useContext, useEffect, useState } from "react";
import styled from "styled-components";
import axios from "axios";
// components
import { GoBack, CartItem, LoadingSpinner } from "../components";
import { Link, useHistory } from "react-router-dom";
// context
import CartContext from "../context/CartContext";
// breakpoints
import { breakpoints } from "../breakpoints";

const StyledCart = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 5%;
    min-height: 45vh;
`;

const StyledCartDiv = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 1rem;
    border: 2px solid #207070;
    box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
    margin-bottom: 1rem;

    ${breakpoints("width", "%", [{ 1024: 70 }])}
`;

const StyledLink = styled(Link)`
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.2rem;
    text-decoration: none;
    color: #207070;
    margin-top: 1rem;

    ${breakpoints("font-size", "rem", [{ 414: 1.4 }])}
`;

const TotalDiv = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
    padding: 0.5rem 0.5rem 0.25rem 0.5rem;
    border: 2px solid #207070;
    box-shadow: 4px 4px 4px rgba(0, 0, 0, 0.25);
    align-self: flex-end;
    margin-bottom: 1rem;

    ${breakpoints("margin-right", "%", [{ 1024: 15 }])}
`;

const TotalText = styled.p`
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.2rem;
    letter-spacing: 1px;
    margin-bottom: 0.25rem;

    ${breakpoints("font-size", "rem", [{ 414: 1.4 }])}
`;

const StyledSpan = styled.span`
    font-weight: 400;
`;

const StyledButton = styled.button`
    background-color: #207070;
    border: 2px solid #207070;
    border-radius: 4px;
    color: #ffffff;
    font-family: "News Cycle", sans-serif;
    font-weight: 700;
    font-size: 1.2rem;
    letter-spacing: 1px;
    padding: 2.5px 10px;
    align-self: flex-end;

    &:hover {
        color: #207070;
        background-color: #ffffff;
        cursor: pointer;
    }

    ${breakpoints("font-size", "rem", [{ 414: 1.4 }, { 768: 1.6 }])}
    ${breakpoints("margin-right", "%", [{ 1024: 15 }])}
`;

const Cart = () => {
    // set up state
    const [shipping, setShipping] = useState(0);
    const [priceTotal, setPriceTotal] = useState(0);
    const [loading, setLoading] = useState(true);
    // set up context
    const { cart, setLocalStorageCart, setOrderId } = useContext(CartContext);
    // set up history
    const history = useHistory();

    useEffect(() => {
        // on component mount, sum shipping and price
        let shippingTotal = 0;
        let priceTotal = 0;
        for (let item of cart) {
            shippingTotal = shippingTotal + +item.shipping;
            priceTotal = priceTotal + +item.price;
        }
        setShipping(shippingTotal.toFixed(2));
        setPriceTotal(priceTotal.toFixed(2));
        setLoading(false);
    }, [cart]);

    const handleClick = () => {
        const checkCart = async () => {
            // when "Checkout" button clicked create order,
            try {
                const res = await axios.post(
                    process.env.REACT_APP_BACKEND_URL + "orders/checkout",
                    cart
                );
                const notAdded = res.data.notAdded;
                // if items not added, abort order, update cart,
                // and send user to cart error page
                if (notAdded.length > 0) {
                    // abort order
                    await axios.delete(
                        process.env.REACT_APP_BACKEND_URL +
                            `orders/order/${res.data.order.id}/abort`
                    );
                    // reset cart to only items that are available
                    for (let id of notAdded) {
                        let newCart = cart;
                        let idx = cart.findIndex(
                            (cartItem) => cartItem.id === id
                        );
                        newCart.splice(idx, 1);
                        setLocalStorageCart(() => newCart);
                    }
                    history.push("/cart/error");
                } else {
                    // if no cart error, set state for order id and send user
                    // to next step of checkout process
                    setOrderId(res.data.order.id);
                    history.push("/checkout");
                }
            } catch (err) {
                console.log("Checkout Error:", err);
                history.push("/cart");
            }
        };
        // check cart for items that are out of stock
        checkCart();
        // if no errors, send user to next step of checkout
        history.push(`/checkout`);
    };

    if (loading)
        return (
            <StyledCart>
                <LoadingSpinner />
            </StyledCart>
        );

    return (
        <StyledCart>
            <GoBack to="/store" />
            <StyledCartDiv>
                {cart.length > 0 ? (
                    cart.map((item, idx) => (
                        <CartItem key={`C${idx}-I${item.id}`} item={item} />
                    ))
                ) : (
                    <>
                        <TotalText>There is nothing in your cart.</TotalText>
                        <StyledLink to={"/store"}>Go add something!</StyledLink>
                    </>
                )}
            </StyledCartDiv>
            {cart.length > 0 && (
                <>
                    <TotalDiv>
                        <TotalText>
                            Shipping: <StyledSpan>${shipping}</StyledSpan>
                        </TotalText>
                        <TotalText>
                            Items: <StyledSpan>${priceTotal}</StyledSpan>
                        </TotalText>
                        <TotalText>
                            Subtotal:{" "}
                            <StyledSpan>
                                ${(+priceTotal + +shipping).toFixed(2)}
                            </StyledSpan>
                        </TotalText>
                    </TotalDiv>
                    <StyledButton onClick={handleClick}>Checkout</StyledButton>
                </>
            )}
        </StyledCart>
    );
};

export default Cart;
