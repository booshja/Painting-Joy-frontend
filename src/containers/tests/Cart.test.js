import { render } from "@testing-library/react";
import Cart from "../Cart";
import CartContext from "../../context/CartContext";
import { BrowserRouter } from "react-router-dom";
import "jest-styled-components";

/** Smoke Test */
it("renders without crashing, nothing in cart", () => {
    render(
        <BrowserRouter>
            <CartContext.Provider
                value={{
                    cart: [],
                    setLocalStorageCart: jest.fn(),
                    setOrderId: jest.fn(),
                }}
            >
                <Cart />
            </CartContext.Provider>
        </BrowserRouter>
    );
});

it("renders without crashing, items in cart", () => {
    render(
        <BrowserRouter>
            <CartContext.Provider
                value={{
                    cart: [1, 2, 3, 4],
                    setLocalStorageCart: jest.fn(),
                    setOrderId: jest.fn(),
                }}
            >
                <Cart />
            </CartContext.Provider>
        </BrowserRouter>
    );
});

/** Snapshot Test */
it("matches snapshot, nothing in cart", () => {
    const { asFragment } = render(
        <BrowserRouter>
            <CartContext.Provider
                value={{
                    cart: [],
                    setLocalStorageCart: jest.fn(),
                    setOrderId: jest.fn(),
                }}
            >
                <Cart />
            </CartContext.Provider>
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

it("matches snapshot, items in cart", () => {
    const { asFragment } = render(
        <BrowserRouter>
            <CartContext.Provider
                value={{
                    cart: [1, 2, 3, 4],
                    setLocalStorageCart: jest.fn(),
                    setOrderId: jest.fn(),
                }}
            >
                <Cart />
            </CartContext.Provider>
        </BrowserRouter>
    );
    expect(asFragment()).toMatchSnapshot();
});

/** RTL Tests */
it("renders correctly, nothing in cart", async () => {
    const { findByText } = render(
        <BrowserRouter>
            <CartContext.Provider
                value={{
                    cart: [],
                    setLocalStorageCart: jest.fn(),
                    setOrderId: jest.fn(),
                }}
            >
                <Cart />
            </CartContext.Provider>
        </BrowserRouter>
    );

    await findByText("Back");
    await findByText("There is nothing in your cart.");
    await findByText("Go add something!");
});

it("renders correctly, items in cart", async () => {
    const { findByText, findByRole } = render(
        <BrowserRouter>
            <CartContext.Provider
                value={{
                    cart: [1, 2, 3, 4],
                    setLocalStorageCart: jest.fn(),
                    setOrderId: jest.fn(),
                }}
            >
                <Cart />
            </CartContext.Provider>
        </BrowserRouter>
    );

    await findByText("Back");
    await findByText("Shipping:");
    await findByText("Items:");
    await findByText("Subtotal:");
    await findByRole("button", { name: "Checkout" });
});
