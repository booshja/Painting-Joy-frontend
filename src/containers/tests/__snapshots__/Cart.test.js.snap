// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot, items in cart 1`] = `
<DocumentFragment>
  .c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem 0;
}

.c4 {
  width: 40%;
  margin-right: 1rem;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  position: relative;
}

.c6 {
  font-family: "Barlow Condensed",sans-serif;
  font-size: 1.2rem;
  font-weight: 700;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 1rem;
}

.c7 {
  font-family: "News Cycle",sans-serif;
  font-size: 1rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 0.5rem;
}

.c8 {
  font-family: "News Cycle",sans-serif;
  font-size: 1rem;
  font-weight: 700;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  color: #207070;
  border: none;
  background-color: transparent;
  padding: 0;
}

.c1 {
  outline: none;
  background: transparent;
  color: #207070;
  font-family: "News Cycle",sans-serif;
  font-size: 1.4rem;
  font-weight: 700;
  cursor: pointer;
  border: none;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin: 0.5rem 0 0.5rem 7.5%;
}

.c1 i {
  margin-right: 0.5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5%;
  min-height: 45vh;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem;
  border: 2px solid #207070;
  box-shadow: 4px 4px 4px rgba(0,0,0,0.25);
  margin-bottom: 1rem;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.5rem 0.5rem 0.25rem 0.5rem;
  border: 2px solid #207070;
  box-shadow: 4px 4px 4px rgba(0,0,0,0.25);
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
  margin-bottom: 1rem;
}

.c10 {
  font-family: "News Cycle",sans-serif;
  font-weight: 700;
  font-size: 1.2rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 0.25rem;
}

.c11 {
  font-weight: 400;
}

.c12 {
  background-color: #207070;
  border: 2px solid #207070;
  border-radius: 4px;
  color: #ffffff;
  font-family: "News Cycle",sans-serif;
  font-weight: 700;
  font-size: 1.2rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  padding: 2.5px 10px;
  -webkit-align-self: flex-end;
  -ms-flex-item-align: end;
  align-self: flex-end;
}

.c12:hover {
  color: #207070;
  background-color: #ffffff;
  cursor: pointer;
}

@media screen and (min-width:414px) {
  .c6 {
    font-size: 1.4rem;
  }
}

@media screen and (min-width:768px) {
  .c6 {
    font-size: 2rem;
  }
}

@media screen and (min-width:1024px) {
  .c6 {
    font-size: 1.6rem;
  }
}

@media screen and (min-width:414px) {
  .c7 {
    font-size: 1.2rem;
  }
}

@media screen and (min-width:768px) {
  .c7 {
    font-size: 1.6rem;
  }
}

@media screen and (min-width:1024px) {
  .c7 {
    font-size: 1.4rem;
  }
}

@media screen and (min-width:414px) {
  .c8 {
    font-size: 1.2rem;
  }
}

@media screen and (min-width:768px) {
  .c8 {
    font-size: 1.6rem;
  }
}

@media screen and (min-width:1024px) {
  .c8 {
    font-size: 1.4rem;
  }
}

@media screen and (min-width:1024px) {
  .c2 {
    width: 70%;
  }
}

@media screen and (min-width:1024px) {
  .c9 {
    margin-right: 15%;
  }
}

@media screen and (min-width:414px) {
  .c10 {
    font-size: 1.4rem;
  }
}

@media screen and (min-width:414px) {
  .c12 {
    font-size: 1.4rem;
  }
}

@media screen and (min-width:768px) {
  .c12 {
    font-size: 1.6rem;
  }
}

@media screen and (min-width:1024px) {
  .c12 {
    margin-right: 15%;
  }
}

<div
    class="c0"
  >
    <button
      class="c1"
    >
      <i
        class="fas fa-long-arrow-alt-left"
      />
      Back
    </button>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <img
          class="c4"
          src="http://localhost:5000/items/item/undefined/image"
        />
        <div
          class="c5"
        >
          <p
            class="c6"
          />
          <p
            class="c7"
          >
            Price: $
          </p>
          <p
            class="c7"
          >
            Shipping: $
          </p>
          <button
            class="c8"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        class="c3"
      >
        <img
          class="c4"
          src="http://localhost:5000/items/item/undefined/image"
        />
        <div
          class="c5"
        >
          <p
            class="c6"
          />
          <p
            class="c7"
          >
            Price: $
          </p>
          <p
            class="c7"
          >
            Shipping: $
          </p>
          <button
            class="c8"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        class="c3"
      >
        <img
          class="c4"
          src="http://localhost:5000/items/item/undefined/image"
        />
        <div
          class="c5"
        >
          <p
            class="c6"
          />
          <p
            class="c7"
          >
            Price: $
          </p>
          <p
            class="c7"
          >
            Shipping: $
          </p>
          <button
            class="c8"
          >
            Remove
          </button>
        </div>
      </div>
      <div
        class="c3"
      >
        <img
          class="c4"
          src="http://localhost:5000/items/item/undefined/image"
        />
        <div
          class="c5"
        >
          <p
            class="c6"
          />
          <p
            class="c7"
          >
            Price: $
          </p>
          <p
            class="c7"
          >
            Shipping: $
          </p>
          <button
            class="c8"
          >
            Remove
          </button>
        </div>
      </div>
    </div>
    <div
      class="c9"
    >
      <p
        class="c10"
      >
        Shipping: 
        <span
          class="c11"
        >
          $NaN
        </span>
      </p>
      <p
        class="c10"
      >
        Items: 
        <span
          class="c11"
        >
          $NaN
        </span>
      </p>
      <p
        class="c10"
      >
        Subtotal: 
        <span
          class="c11"
        >
          $NaN
        </span>
      </p>
    </div>
    <button
      class="c12"
    >
      Checkout
    </button>
  </div>
</DocumentFragment>
`;

exports[`matches snapshot, nothing in cart 1`] = `
<DocumentFragment>
  .c1 {
  outline: none;
  background: transparent;
  color: #207070;
  font-family: "News Cycle",sans-serif;
  font-size: 1.4rem;
  font-weight: 700;
  cursor: pointer;
  border: none;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin: 0.5rem 0 0.5rem 7.5%;
}

.c1 i {
  margin-right: 0.5rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 5%;
  min-height: 45vh;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 1rem;
  border: 2px solid #207070;
  box-shadow: 4px 4px 4px rgba(0,0,0,0.25);
  margin-bottom: 1rem;
}

.c4 {
  font-family: "News Cycle",sans-serif;
  font-weight: 700;
  font-size: 1.2rem;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: #207070;
  margin-top: 1rem;
}

.c3 {
  font-family: "News Cycle",sans-serif;
  font-weight: 700;
  font-size: 1.2rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 0.25rem;
}

@media screen and (min-width:1024px) {
  .c2 {
    width: 70%;
  }
}

@media screen and (min-width:414px) {
  .c4 {
    font-size: 1.4rem;
  }
}

@media screen and (min-width:414px) {
  .c3 {
    font-size: 1.4rem;
  }
}

<div
    class="c0"
  >
    <button
      class="c1"
    >
      <i
        class="fas fa-long-arrow-alt-left"
      />
      Back
    </button>
    <div
      class="c2"
    >
      <p
        class="c3"
      >
        There is nothing in your cart.
      </p>
      <a
        class="c4"
        href="/store"
      >
        Go add something!
      </a>
    </div>
  </div>
</DocumentFragment>
`;
